cassandra-analytics-spark-converter/src/main/java/org/apache/cassandra/spark/data/converter/types/complex/SparkTuple.java

```diff
@@ -0,0 +1,141 @@
+import org.apache.cassandra.spark.data.converter.types.SparkType;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow;
+import org.apache.spark.sql.catalyst.expressions.GenericRowWithSchema;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.StructField;
+
+public class SparkTuple implements CollectionFeatures
+{
+    private final SparkSqlTypeConverter converter;
+    private final CqlField.CqlTuple tuple;
+
+    public SparkTuple(SparkSqlTypeConverter converter, CqlField.CqlTuple tuple)
+    {
+        this.converter = converter;
+        this.tuple = tuple;
+    }
+
+    @Override
+    public DataType dataType(BigNumberConfig bigNumberConfig)
+    {
+        return DataTypes.createStructType(IntStream.range(0, size())
+                                                   .mapToObj(index -> DataTypes.createStructField(
+                                                   Integer.toString(index),
+                                                   sparkType(index).dataType(bigNumberConfig),
+                                                   true))
+                                                   .toArray(StructField[]::new)
+        );
+    }
+
+    public CqlField.CqlCollection collection()
+    {
+        return tuple;
+    }
+
+    public SparkSqlTypeConverter converter()
+    {
+        return converter;
+    }
+
+    @Override
+    public Object toSparkSqlType(Object value, boolean isFrozen)
+    {
+        if (value instanceof ByteBuffer)
+        {
+            // Need to deserialize first, e.g. if tuple is frozen inside collections
+            return toSparkSqlType(tuple.deserializeToJavaType((ByteBuffer) value), isFrozen);
+        }
+        else
+        {
+            if (!(value instanceof Object[]))
+            {
+                throw new IllegalArgumentException("Expected Object[] or raw ByteBuffer for Tuple type");
+            }
+            Object[] array = (Object[]) value;
+            for (int index = 0; index < array.length; index++)
+            {
+                array[index] = array[index] == null ? null : sparkType(index).toSparkSqlType(array[index], isFrozen);
+            }
+            return new GenericInternalRow(array);
+        }
+    }
+
+    @Override
+    public Object sparkSqlRowValue(GenericInternalRow row, int position)
+    {
+        final InternalRow tupleStruct = row.getStruct(position, size());
+        return IntStream.range(0, size())
+                        .boxed()
+                        .map(index -> sparkType(index).toTestRowType(tupleStruct.get(index, sparkType(index).dataType())))
+                        .toArray();
+    }
+
+    @Override
+    public Object sparkSqlRowValue(Row row, int position)
+    {
+        final Row tupleStruct = row.getStruct(position);
```

Comment:
NIT unnecessary final modifier
```suggestion
        Row tupleStruct = row.getStruct(position);
```